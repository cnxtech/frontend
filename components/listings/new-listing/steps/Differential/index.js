import React, { Component } from 'react'
import { Formik, Field } from 'formik'

import Input from '@emcasa/ui-dom/components/Input'
import Row from '@emcasa/ui-dom/components/Row'
import Col from '@emcasa/ui-dom/components/Col'
import Text from '@emcasa/ui-dom/components/Text'
import {withBreakpoint} from '@emcasa/ui-dom/components/Text'
import NavButtons from 'components/listings/new-listing/shared/NavButtons'
import { getAddressInput } from 'lib/address'
import { estimatePrice, getPricingInput } from 'lib/listings/pricing'
import { getUser, hasPhoneNumber } from 'components/listings/new-listing/lib/auth'

class Differential extends Component {
  constructor(props) {
    super(props)
    this.previousStep = this.previousStep.bind(this)
    this.updateStateFromProps = this.updateStateFromProps.bind(this)

    this.checkUserInfo = this.checkUserInfo.bind(this)
    this.estimatePrice = this.estimatePrice.bind(this)
    this.submit = this.submit.bind(this)

    this.textInput = React.createRef()
  }

  state = {
    loading: false,
    error: null,
    text: null
  }

  componentDidMount() {
    this.updateStateFromProps(this.props)
    if (this.textInput.current && !this.props.isMobile) {
      this.textInput.current.focus()
    }
  }
  componentWillReceiveProps(props) {
    this.updateStateFromProps(props)
  }

  updateStateFromProps(props) {
    const { differential } = props
    if (differential) {
      this.setState({
        text: differential.text
      })
    }
  }

  previousStep() {
    const { navigateTo } = this.props
    navigateTo('garage')
  }

  async checkUserInfo() {
    const { updateDifferential, user } = this.props
    updateDifferential({text: this.state.text})
    const authenticated = user && user.authenticated

    // Old site auth
    if (authenticated) {
      const { updatePhone, updatePersonal } = this.props
      try {
        const userInfo = await getUser(user.id, updatePhone, updatePersonal)
        await this.estimatePrice(userInfo)
        return
      } catch (e) {
        this.setState({
          loading: false,
          error: 'Ocorreu um erro. Por favor, tente novamente.'
        })
        return
      }
    }

    // User has already input phone number
    const { phone } = this.props
    if (hasPhoneNumber(phone)) {
      const { personal } = this.props
      if (personal && personal.name) {
        await this.estimatePrice()
        return
      } else {
        this.props.navigateTo('personal')
        return
      }
    }

    this.props.navigateTo('phone')
  }

  async estimatePrice(userInfo) {
    // Prepare input
    const { personal, homeDetails, rooms, garage, location } = this.props
    const addressInput = getAddressInput(location.addressData)
    const pricingInput = getPricingInput(addressInput, homeDetails, rooms, garage, personal, userInfo)

    // Run mutation
    const response = await estimatePrice(apolloClient, pricingInput)
    if (response.error) {
      this.setState({
        loading: false,
        error: response.error
      })
    }

    // Handle result
    if (response.result) {
      const { suggestedPrice, userPrice } = response.result
      const { navigateTo, updatePricing, updateDifferential, pricing } = this.props
      updatePricing({
        ...pricing,
        suggestedPrice,
        userPrice
      })
      updateDifferential({text: this.state.text})
      this.setState({
        loading: false,
        error: null
      })
      navigateTo('pricing')
    }
  }

  submit() {
    this.setState({loading: true})
    this.checkUserInfo()
  }

  render() {
    const { differential } = this.props
    let text
    if (differential) {
      text = differential.text
    }
    return (
      <div ref={this.props.hostRef}>
        <Row justifyContent="center" p={4} pt={0}>
          <Col width={[1, 1/2]}>
            <Formik
              initialValues={{
                text: text
              }}
              isInitialValid={() => true}
              render={({setFieldTouched, setFieldValue}) => (
                <>
                  <Text
                    fontSize="large"
                    fontWeight="bold"
                    textAlign="center">
                    Seu imóvel tem algum diferencial?
                  </Text>
                  <Text color="grey">Conte pra gente algum diferencial do seu imóvel.</Text>
                  <Row mb={4}>
                    <Col width={1}>
                      <Field
                        name="text"
                        render={() => (
                          <Input
                            area
                            hideLabelView
                            hideErrorView
                            ref={this.textInput}
                            placeholder="Diferenciais do imóvel"
                            defaultValue={text}
                            style={{height: 150}}
                            onChange={(e) => {
                              const { value } = e.target
                              setFieldValue('text', value)
                              setFieldTouched('text')
                              this.setState({text: value})
                            }}
                          />
                        )}/>
                    </Col>
                  </Row>
                  <Text color="red">{this.state.error}</Text>
                  <NavButtons
                    disableEnterToSubmit
                    submitEnabled={!this.state.loading}
                    loading={this.state.loading}
                    previousStep={this.previousStep}
                    onSubmit={this.submit}
                  />
                </>
              )}
            />
          </Col>
        </Row>
      </div>
    )
  }
}

export default withBreakpoint()(Differential)
