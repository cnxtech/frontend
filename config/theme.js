import theme from '@emcasa/ui'
import {fas} from '@fortawesome/pro-solid-svg-icons'
import {fal} from '@fortawesome/pro-light-svg-icons'
import icons from './icons'

export default {
  ...theme,
  icons: {emcasa: icons, default: fas, light: fal}
}
