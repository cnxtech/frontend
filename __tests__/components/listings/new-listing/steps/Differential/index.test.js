import React from 'react'
import Differential from 'components/listings/new-listing/steps/Differential'
import renderer from 'react-test-renderer'
import 'jest-styled-components'
import theme from '@emcasa/ui'
import { ThemeProvider } from 'styled-components'

it('renders correctly', () => {
  const tree = renderer.create(
    <ThemeProvider theme={theme}>
      <Differential />
    </ThemeProvider>).toJSON()
  expect(tree).toMatchSnapshot()
})
