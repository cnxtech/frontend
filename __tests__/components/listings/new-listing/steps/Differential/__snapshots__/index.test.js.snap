// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
.c2 {
  -webkit-font-smoothing: antialiased;
  font-family: FaktSoftPro-Normal;
  color: #3c4858;
  font-size: 24px;
  font-weight: bold;
  text-align: center;
}

.c3 {
  -webkit-font-smoothing: antialiased;
  font-family: FaktSoftPro-Normal;
  color: #979797;
  font-size: 18px;
}

.c7 {
  -webkit-font-smoothing: antialiased;
  font-family: FaktSoftPro-Normal;
  color: #ff4949;
  font-size: 18px;
}

.c6 {
  -webkit-font-smoothing: antialiased;
  display: block;
  box-sizing: border-box;
  width: 100%;
  outline: none;
  border-radius: 4px;
  border: 1px solid;
  border-color: #e0e6ed;
  padding: 10px 10px 10px 10px;
  height: 60px;
  color: #3c4858;
  font-family: FaktSoftPro-Normal;
  font-size: 18px;
  -webkit-letter-spacing: 0px;
  -moz-letter-spacing: 0px;
  -ms-letter-spacing: 0px;
  letter-spacing: 0px;
}

.c6:focus {
  border-radius: 4px;
  border: 1px solid;
  border-color: #1fb6ff;
  padding: 10px 10px 10px 10px;
  height: 60px;
}

.c6::-webkit-input-placeholder {
  color: #8492a6;
}

.c6::-moz-placeholder {
  color: #8492a6;
}

.c6:-ms-input-placeholder {
  color: #8492a6;
}

.c6::placeholder {
  color: #8492a6;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  padding: 20px;
  padding-top: 0px;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-bottom: 20px;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin-top: 20px;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c1 {
  width: 100%;
}

.c5 {
  width: 100%;
}

.c9 {
  width: 41.66666666666667%;
}

.c10 {
  outline: none;
  display: block;
  cursor: pointer;
  padding: 0 20px 0 20px;
  border-color: #e0e6ed;
  background-color: #ffffff;
  border-radius: 4px;
  border: 1px solid;
  width: 100%;
  height: 60px;
  color: #3c4858;
  font-family: FaktSoftPro-Normal;
  font-size: 18px;
  -webkit-font-smoothing: antialiased;
  border-color: #e0e6ed;
}

.c10 :hover {
  background-color: #ffffff;
}

.c11 {
  outline: none;
  display: block;
  cursor: pointer;
  padding: 0 20px 0 20px;
  border-color: #f50057;
  background-color: #f50057;
  border-radius: 4px;
  border: 1px solid;
  width: 100%;
  height: 60px;
  color: #ffffff;
  font-family: FaktSoftPro-Normal;
  font-size: 18px;
  -webkit-font-smoothing: antialiased;
  border-color: #e0e6ed;
}

.c11 :hover {
  background-color: #f50057;
}

@media screen and (min-width:780px) {
  .c1 {
    width: 50%;
  }
}

<div>
  <div
    className="c0"
  >
    <div
      className="c1"
      width={
        Array [
          1,
          0.5,
        ]
      }
    >
      <p
        className="c2"
        color="dark"
        fontSize="large"
        fontWeight="bold"
      >
        Seu imóvel tem algum diferencial?
      </p>
      <p
        className="c3"
        color="grey"
        fontSize="default"
      >
        Conte pra gente algum diferencial do seu imóvel.
      </p>
      <div
        className="c4"
      >
        <div
          className="c5"
          width={1}
        >
          <div
            className=""
          >
            <textarea
              className="c6"
              fontSize="default"
              height="tall"
              letterSpacing="default"
              onChange={[Function]}
              placeholder="Diferenciais do imóvel"
              style={
                Object {
                  "height": 150,
                }
              }
            />
          </div>
        </div>
      </div>
      <p
        className="c7"
        color="red"
        fontSize="default"
      />
      <div
        className="c8"
      >
        <div
          className="c9"
          width={0.4166666666666667}
        >
          <button
            className="c10"
            fontSize="default"
            height="tall"
            onClick={[Function]}
          >
            Voltar
          </button>
        </div>
        <div
          className="c9"
          width={0.4166666666666667}
        >
          <button
            className="c11"
            disabled={false}
            fontSize="default"
            height="tall"
            onClick={[Function]}
          >
            Avançar
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
